dartagnan/src/main/java/com/dat3m/dartagnan/program/memory/Memory.java

```diff
@@ -26,36 +35,68 @@ public Location getLocationForAddress(Address address){
+        				fmgr.getIntegerFormulaManager().greaterThan(fmgr.getBitvectorFormulaManager().toIntegerFormula((BitvectorFormula)add.toIntFormula(ctx), false), fmgr.getIntegerFormulaManager().makeNumber(BigInteger.ZERO)) :
+        				fmgr.getIntegerFormulaManager().greaterThan((IntegerFormula)add.toIntFormula(ctx), fmgr.getIntegerFormulaManager().makeNumber(BigInteger.ZERO)));
         	}
         }
         
-        return ctx.mkAnd(enc, ctx.mkDistinct(getAllAddresses().stream().map(a -> a.toZ3Int(ctx).isBV() ? ctx.mkBV2Int((BitVecExpr) a.toZ3Int(ctx), false) : a.toZ3Int(ctx)).toArray(Expr[]::new)));
+        BooleanFormula distinct = getAllAddresses().size() > 1 ?
+        		fmgr.getIntegerFormulaManager().distinct(getAllAddresses().stream()
+                		.map(a -> convertToIntegerFormula(a.toIntFormula(ctx), ctx))
+                		.collect(Collectors.toList())) : 
+                bmgr.makeTrue();
+        
+        return bmgr.and(enc, distinct);        	
     }
 
+    private IntegerFormula convertToIntegerFormula(Formula f, SolverContext ctx) {
+    	return f instanceof BitvectorFormula ? 
+    			ctx.getFormulaManager().getBitvectorFormulaManager().toIntegerFormula((BitvectorFormula) f, false) : 
+    			(IntegerFormula)f;
+    }
+    
     // Assigns each Address a fixed memory address.
-    public BoolExpr fixedMemoryEncoding(Context ctx) {
-        BoolExpr[] addrExprs = getAllAddresses().stream().filter(x -> !x.hasConstantValue())
-        		.map(add -> {
-                    Expr e1 = add.toZ3Int(ctx);
-                    e1 = e1.isBV() ? ctx.mkBV2Int(e1, false) : e1;
-                    return ctx.mkEq(e1, ctx.mkInt(add.getValue().intValue()));
-                }).toArray(BoolExpr[]::new);
-        return ctx.mkAnd(addrExprs);
+    public BooleanFormula fixedMemoryEncoding(SolverContext ctx) {
+        FormulaManager fmgr = ctx.getFormulaManager();
+
+        boolean bv = getAllAddresses().iterator().next().toIntFormula(ctx) instanceof BitvectorFormula;
+    	BooleanFormula[] addrExprs;
+    	if(bv) {
+        	addrExprs = getAllAddresses().stream().filter(x -> !x.hasConstantValue())
+            		.map(add -> {
+                        Formula e1 = add.toIntFormula(ctx);
+                        fmgr.getBitvectorFormulaManager().toIntegerFormula((BitvectorFormula) e1, false);
+						return fmgr.getIntegerFormulaManager().equal((IntegerFormula) e1, fmgr.getIntegerFormulaManager().makeNumber(add.getValue().intValue()));})
+            		.toArray(BooleanFormula[]::new);
+    	} else {
+        	addrExprs = getAllAddresses().stream().filter(x -> !x.hasConstantValue())
+            		.map(add -> fmgr.getIntegerFormulaManager().equal((IntegerFormula) add.toIntFormula(ctx), fmgr.getIntegerFormulaManager().makeNumber(add.getValue().intValue())))
+            		.toArray(BooleanFormula[]::new);
+    	}
+        return fmgr.getBooleanFormulaManager().and(addrExprs);
```

Comment:
- Refactor fmgr.getIntegerFormulaManager into local variable
- Use convertToIntegerFormula and get rid of case distinction (there is not much performance gain in checking for bv once vs. in every iteration).
- The bv check fails if there are no addresses (should rarely happen but may be the case for sequential programs)